Overhaul log loading

CyberShadow opened the pull request:
Unify the channel/query log loading code.

Change algorithm to use the same filename generation algorithm as the
log-creation code.

Allow loading lines from more than one log.

Allow loading logs from all supported date formats.

Fix newline parsing being dependent on compilation target OS.

CyberShadow committed:
Overhaul log loading

Unify the channel/query log loading code.

Change algorithm to use the same filename generation algorithm as the
log-creation code.

Allow loading lines from more than one log.

Allow loading logs from all supported date formats.

Fix newline parsing being dependent on compilation target OS.

Fixes #1281, fixes #1381.

un1versal commented:
if https://github.com/kvirc/KVIrc/issues/1710 ends up being also ~~fixed~~ implemented would be nice :)

wodim commented:
If this works, you are our hero. This was hairy and buggy as fuck, and many people requested it

un1versal commented on src/kvirc/ui/KviWindow.cpp:
```diff
@@ -1380,3 +1390,114 @@ KviIrcContext * KviWindow::context()
+				{
+					vLines[uCurLines++] = QString(list.at(--uCount));
+					if (uCurLines == uMaxLines)
+						goto enough;
```
goto (this reminds me of ZX-spectrum and basic programming) ahh fond memories

CyberShadow committed:
Move log loading into KviChannel/QueryWindow constructor

- Fixes query log not loading when the query window is opened by user
  action (e.g. double-clicking a nick in the member list).

- Fixes the loaded log lines being once again saved back to the log.

CyberShadow committed:
Propagate correct timestamp to logging code

Fixes incorrect logging of server-time timestamps.

CyberShadow committed:
Load timestamps and message type from logs

Parse the logged timestamps and message type, if logging them was
enabled.

CyberShadow committed:
Add KVI_OUT_LOG

To be used for log meta-information (or for when message types were not
logged).

Uses the existing log icon (which incidentally wasn't used by any
message type yet).

CyberShadow committed:
Tweak log session marker messages

Log the timestamp in the same way that all other messages' timestamps
are logged, so that they can be loaded correctly.

Use the new KVI_OUT_LOG message type.

CyberShadow committed:
Log milliseconds in ISO timestamps

CyberShadow committed:
Raise limits on log loading options

CyberShadow committed:
Work around QDateTime::fromString bug

CyberShadow commented:
OK, I think this is ready for review & testing. See the individual commits' commit messages for details on changes done.

un1versal commented:
so is this ready to go? We had a few log related issues if this fix resolves any of them
see https://github.com/kvirc/KVIrc/milestone/5

Would you mind adding **fixes #issuenr** to your commit messages for any of those reports you see fixed by this, and Ill take care of the reports relabeling.

CyberShadow commented:
@un1versal Added fixed issue references to commit message.

CyberShadow commented:
> if #1710 ends up being also fixed implemented would be nice :)

This one is orthogonal to the fixes here.

un1versal commented:
> @un1versal Added fixed issue references to commit message.

Thank you... Its preferred to close tickets automatically, but, if you really forget or are too busy to go around adding these to commit msg, if you reference them in a post here, when PR is merged, I can go and close the related issues and refer to the fix for posterity.

> This one is orthogonal to the fixes here.

Yes I see #2054 staticfocx will be pleased.

nice going, this logging thing is important for national security reasons `</sarcasm>` hahaha

un1versal commented:
Im not sure if its a placebo effect or not, but when opening logviewer, it feels faster somehow.

Linux test :+1: 
Windows test: :+1: 

if none has anything else to say, Ill merge this after a short break for commercials.
